src/map.rs

```diff
@@ -320,89 +417,641 @@ where
+            let sc = self.size_ctl.load(Ordering::SeqCst);
+            if (count as isize) < sc {
+                // we're not at the next resize point yet
+                break;
+            }
+
+            let table = self.table.load(Ordering::SeqCst, guard);
+            if table.is_null() {
+                // table will be initalized by another thread anyway
+                break;
+            }
+
+            // safety: table is only dropped on the next epoch change after it is swapped to null.
+            // we read it as not null, so it must not be dropped until a subsequent epoch. since we
+            // hold a Guard, we know that the current epoch will persist, and that our reference
+            // will therefore remain valid.
+            let n = unsafe { table.deref() }.len();
+            if n >= MAXIMUM_CAPACITY {
+                // can't resize any more anyway
+                break;
+            }
+
+            let rs = Self::resize_stamp(n) << RESIZE_STAMP_SHIFT;
+            if sc < 0 {
+                // ongoing resize! can we join the resize transfer?
+                if sc == rs + MAX_RESIZERS || sc == rs + 1 {
+                    break;
+                }
+                let nt = self.next_table.load(Ordering::SeqCst, guard);
+                if nt.is_null() {
+                    break;
+                }
+                if self.transfer_index.load(Ordering::SeqCst) <= 0 {
+                    break;
+                }
+
+                // try to join!
+                if self.size_ctl.compare_and_swap(sc, sc + 1, Ordering::SeqCst) == sc {
+                    self.transfer(table, nt, guard);
+                }
+            } else if self.size_ctl.compare_and_swap(sc, rs + 2, Ordering::SeqCst) == sc {
+                // a resize is needed, but has not yet started
+                // TODO: figure out why this is rs + 2, not just rs
+                // NOTE: this also applies to `try_presize`
+                self.transfer(table, Shared::null(), guard);
+            }
+
+            // another resize may be needed!
+            count = self.count.load(Ordering::SeqCst);
+        }
+    }
+
+    #[inline]
+    /// Tries to reserve capacity for at least `additional` more elements to
+    /// be inserted in the `HashMap`. The collection may reserve more space to
+    /// avoid frequent reallocations.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use flurry::HashMap;
+    ///
+    /// let map: HashMap<&str, i32> = HashMap::new();
+    ///
+    /// map.pin().reserve(10);
+    /// ```
+    ///
+    /// # Notes
+    ///
+    /// Reserving does not panic in flurry. If the new size is invalid, no
+    /// reallocation takes place.
+    pub fn reserve(&self, additional: usize, guard: &Guard) {
+        self.check_guard(guard);
+        let absolute = self.len() + additional;
+        self.try_presize(absolute, guard);
+    }
+
+    fn init_table<'g>(&'g self, guard: &'g Guard) -> Shared<'g, Table<K, V>> {
```

Comment:
This is in the `K: Clone` `impl` block, but it doesn't handle entries, initializes the table with `null` bins and doesn't call `transfer` or related methods. Why does it need `Clone` on the key type?

Reply:
Good catch! Fixed.